### ISSUE TITLE
AutoDetect throws Security.Authentication.AuthenticationException if server supports TLSv1 only

### ISSUE THREAD

**Issue by dmansurov83** [robinrodricks/FluentFTP#1289]
**FTP Server OS:** Unix

**FTP Server Type:** ProFTPD

**Client Computer OS:** Windows 

**FluentFTP Version:** 47.0.0

**Framework:** Framework 4.6.2

I'm trying to call AutoDetect, but it throws an exception:
System.Security.Authentication.AuthenticationException: A call to SSPI failed, see inner exception. ---> System.ComponentModel.Win32Exception: The client and server cannot communicate, because they do not possess a common algorithm

**Logs :**

```

# AutoDetect(False, False)
2023-07-19 06:51:40.2089| INFO|Connecting to ftp://ftp.***.**

# Connect(False)
Status:   FluentFTP 47.0.0.0(.NET Framework 4.6.2)
Status:   Connecting to IP #1= ***:21
Status:   Waiting for a response
Response: 220 ProFTPD 1.3.5b Server (SNPI) [::ffff:**.**.**.**] [738719.286d]
Status:   Detected FTP server: ProFTPD
Command:  AUTH TLS
Status:   Waiting for response to: AUTH TLS
Response: 234 AUTH TLS successful [77ms]
Status:   Closing/Disposing FtpSocketStream(control connection)
Error:    FTPS Authentication failed, lib = .NET SslStream
Error:    A call to SSPI failed, see inner exception
```

Looks like this is because of the protocols priority:
``` 
    private static List<SslProtocols> DefaultProtocolPriority = new List<SslProtocols>()
    {
      SslProtocols.Tls11 | SslProtocols.Tls12,
      SslProtocols.Default
    };
```
Am I right? Is any way to make profile autodetection work?

**Comment by FanDjango** [comment]
Can you provide the tls.conf file of the server? I am trying to set up such a szenario. Mine fails too, but with a different exception.

Can you "Connect" manually, if so - with which code/parms?

**Comment by dmansurov83** [comment]
> Can you provide the tls.conf file of the server?

No, server isn't manages by us and I don't know the exact server config.

> Can you "Connect" manually, if so - with which code/parms?
I'm trying to use the manual profile like:
``` 
var activeProfile = new FtpProfile()
					{
						Host = _ftpClient.Host,
						Credentials = _ftpClient.Credentials,
						Encoding = Encoding.UTF8,
						Encryption = FtpEncryptionMode.Explicit,
						Protocols = SslProtocols.Tls,
					};
```
And the exception is gone. But another error happens:
``` 
# OpenPassiveDataStreamAsync(PASV, "MLSD /", 0)
Command:  PASV
Status:   Waiting for response to: PASV
Response: 227 Entering Passive Mode (212,129,60,98,161,21). [109ms]
Status:   Connecting to IP #1= ***:41237
Command:  MLSD /
Status:   Waiting for response to: MLSD /
Response: 150 Opening BINARY mode data connection for MLSD [109ms]
Status:   FTPS authentication successful, lib = .NET SslStream, cipher suite = Tls (Aes256, 44550, 256) [225ms]
+---------------------------------------+
-----------------------------------------
Status:   Closing/Disposing FtpSocketStream(data connection)

# CloseDataStream()
Status:   Waiting for response to: MLSD /
Response: 425 Unable to build data connection: Operation not permitted [345ms]
Status:   Closing/Disposing FtpSocketStream(data connection)
Command:  QUIT
Status:   Waiting for response to: QUIT
Response: 550 MLSD: Operation not permitted [<1ms]
Status:   Closing/Disposing FtpSocketStream(control connection)
```
FluentFTP.Exceptions.FtpCommandException: Code: 425 Message: Unable to build data connection: Operation not permitted.

BTW, The same profile but with the `Encryption = FtpEncryptionMode.None` works fine.

**Comment by FanDjango** [comment]
> The same profile but with the Encryption = FtpEncryptionMode.None works fine.

Yes, then there is **no TLS at all**. I am surprised the server allows it. But then, if it is only supporting TLS 1.1 anyway, it isn't a really very secure server.

**Comment by FanDjango** [comment]
> Code: 425 Message: Unable to build data connection: Operation not permitted.

If you google that one in comjunction with "ProFTPD", you get a lot of probable causes. All of them need you to modify the configuration of the server, which you said, you cannot do.

**Comment by FanDjango** [comment]
So: Here are some suggestions:

1. If you cannot find a manual connect that works (**with encryption**), AutoDetect will not find a suitable profile either. Turning off encryption isn't fair, but maybe it will be your only way, **unless**..

2. Consider using the FluentFTP.GnuTLS and manually finding the right "**special** parameters to make this connection work - there are **many** more that can be set (SslStream only lets you select protocols). Also set the logging to be very detailed and one will get (possible) many more hints, where exactly the Ssl Auth is failiing (at what stage etc.).

**Comment by FanDjango** [comment]
Also, for SslStream, the internal logic is by now quite convoluted. You should **always** use `Protocols = SslProtocols.None` to get the highest encryption level the server supports - opting downwards is really not reliable anymore.

It is my suspicion that SslStream simply does not really work with TLS (1.1) anymore in many szeanarios.

Can you actually connect to this server manually **using some client like FileZilla** from this Windows machine? Can you then turn on debug logging in the FileZilla client? It uses GnuTLS, so then we could see the negotiated suites, and then do the same with FluentFTP.GnuTLS.

**Comment by dmansurov83** [comment]
> If you cannot find a manual connect that works (with encryption), AutoDetect will not find a suitable profile either. Turning off encryption isn't fair, but maybe it will be your only way, unless.

I would like to have a working `autodetect` becausewe need to connect to a huge number of different FTP servers. To me DefaultProtocolPriority+DefaultEncryptionPriority should be configurable in order to have the possibility to override those. Also a custom handling of the exception (`IsFtpsFailure`) could make this part of code very flexible.

We used 29.0.1 version a lot of time without updates and there the policy was - `No encryption` first, and it is acceptable to me. But for sure, I would like to have TLS working.

**Comment by FanDjango** [comment]
> and there the policy was - No encryption first

I see. That was before my time...

Ok, here is a _general_ suggestion, as a workaround for now:

Do your AutoDetect, catching the exception `System.Security.Authentication.AuthenticationException: A call to SSPI failed, see inner exception. ---> System.ComponentModel.Win32Exception: The client and server cannot communicate, because they do not possess a common algorithm`.

Not having a common algorithm is a terminal condition -> **cannot** use Encryption. In this case, try again in **your own code** with a manual connect and `Encryption.None`.

Meanwhile I will have a look at the current code to see if this can be improved somehow.

**Comment by FanDjango** [comment]
#1291 Should fix this together with #1290 

Can you test current master?

**Comment by FanDjango** [comment]
> To me DefaultProtocolPriority+DefaultEncryptionPriority should be configurable in order to have the possibility to override those. Also a custom handling of the exception (IsFtpsFailure) could make this part of code very flexible.

I see, that would be a possible next step, I suppose.

Do you **really** have any servers out there that need something **other** than `Tls11|Tls12` **or** `no-encryption`? Honestly, what is there to detect at all? Just two tries? Right?

> there the policy was - No encryption first

I see. But I do think changing the order was according to the best interests of most users.

> Also a custom handling of the exception (IsFtpsFailure) could make this part of code very flexible.

Will do some more thinking about this.

**Comment by dmansurov83** [comment]
> Can you test current master?

Works fine now. The profile auto-detected:
```
	Host = "***.*****.***",
	Credentials = new NetworkCredential("****", "********"),
	Encryption = FtpEncryptionMode.None,
	Protocols = SslProtocols.Tls11 | SslProtocols.Tls12, 
	DataConnection = FtpDataConnectionType.PASV,
	Encoding = System.Text.UTF8Encoding,
```
> Do you really have any servers out there that need something other than Tls11|Tls12 or no-encryption? Honestly, what is there to detect at all? Just two tries? Right?

Yep, Looks like nothing to detect. But you have no idea what junk we sometimes have to work with as a data source). 
As to me, from the library's point of view, it would be polite to let the user manage this policy. Even if 99% of users are satisfied with the default one. In any case, these are just my thoughts, and it's up to you to decide how to further develop the project))

**Comment by dmansurov83** [comment]
How soon can I get a fix in nuget? I'd drop crutches in my code then.

**Comment by FanDjango** [comment]
> How soon can I get a fix in nuget? I'd drop crutches in my code then.

I notified @robinrodricks  - but we are actually still debating how to go a bit further on this (**or not**). So hold your horses for a while.

**Comment by FanDjango** [comment]
See #1300

**Comment by FanDjango** [comment]
This is now released as Nuget **47.1.0**. Say thanks to @robinrodricks - you can even configure it, there are new parameters to steer the way it works, all the bells and whistles.
